nmigen/vendor/altera.py

```diff
@@ -137,6 +137,7 @@ def get_input(self, pin, port, attrs, invert):
                 m.d.comb += pin.i[bit].eq(self._invert_if(invert, ff_i[bit]))
             elif pin.xdr == 1:
                 m.submodules["{}_dff_i_{}".format(pin.name, bit)] += Instance("dff",
+                    a_useioff=True,
```

Comment:
Does this actually work?

Reply:
I have no way of testing SDR, let alone this specifically, and I don't have an oscilloscope to check that SDR output is sane, so probably not.

Reply:
Does this synthesize without complaints from Quartus? What if you mistype the attribute on purpose, does it complain then?

Reply:
> Does this synthesize without complaints from Quartus?
Quartus appears to recognise but ignore the attribute; I had to consult the manual to find out that what Quartus expects is not e.g.
```verilog
module top(input clk50_0__io, input i, output o);
wire ff_o;
(* useioff = 1 *) dff foo (.d(i), .clk(clk50_0__io), .clrn(1'b1), .prn(1'b1), .q(ff_o));
alt_outbuf bar (.i(ff_o), .o(o));
endmodule
```
but
```verilog
module top(input clk50_0__io, (* useioff = 1 *) input i, output o);
wire ff_o;
dff foo (.d(i), .clk(clk50_0__io), .clrn(1'b1), .prn(1'b1), .q(ff_o));
alt_outbuf bar (.i(ff_o), .o(o));
endmodule
```
which gives you
```
Info (176235): Finished register packing
    Extra Info (176218): Packed 1 registers into blocks of type I/O input buffer
```
in the compiler output.

I'm not sure how to express that the top-level arguments of a module/fragment need attributes though.

> What if you mistype the attribute on purpose, does it complain then?
`Warning (10335): Unrecognized synthesis attribute "useiofff" at top.v(5) File: //wsl$/Ubuntu-18.04/home/lofty/de_10/build/top.v Line: 5`

Reply:
Try `pin.i.attrs["useioff"] = "1".

Reply:
Like 15ac723?